File: БР1.1/Полярус_Павел/labs/lab23/notification-service/src/consumer.ts
import amqp from 'amqplib';

const EXCHANGE = 'jobby';
const QUEUES = {
  created: 'q.application.created',
  status: 'q.application.status',
};

export async function startConsumer(): Promise<void> {
  const conn = await amqp.connect(process.env.RABBITMQ_URL!);
  const channel = await conn.createChannel();

  await channel.assertExchange(EXCHANGE, 'topic', { durable: true });

  await channel.assertQueue(QUEUES.created, { durable: true });
  await channel.bindQueue(QUEUES.created, EXCHANGE, 'application.created');

  await channel.assertQueue(QUEUES.status, { durable: true });
  await channel.bindQueue(QUEUES.status, EXCHANGE, 'application.status_changed');

  channel.prefetch(1);

  channel.consume(QUEUES.created, (msg) => {
    if (!msg) return;
    const payload = JSON.parse(msg.content.toString());
    console.log('[notification] application.created:', JSON.stringify(payload, null, 2));
    channel.ack(msg);
  });

  channel.consume(QUEUES.status, (msg) => {
    if (!msg) return;
    const payload = JSON.parse(msg.content.toString());
    console.log('[notification] application.status_changed:', JSON.stringify(payload, null, 2));
    channel.ack(msg);
  });

  console.log('notification-service listening on queues:', Object.values(QUEUES).join(', '));
}

File: БР1.2/Шестак_Богдан/labs/lab3/resume-service/app/consumer.py
import aio_pika
import json
import os
import logging
from sqlalchemy.orm import Session
from app.database import SessionLocal
from app.models import Resume

RABBITMQ_URL = os.getenv("RABBITMQ_URL", "amqp://guest:guest@rabbitmq:5672/")
EXCHANGE_NAME = "job_platform"

logger = logging.getLogger(__name__)


async def handle_application_created(payload: dict):
    """
    Fired when a new application is created.
    resume-service can mark the resume as actively used.
    """
    resume_id = payload.get("resume_id")
    vacancy_id = payload.get("vacancy_id")
    logger.info(
        f"[resume-service] Resume {resume_id} was used in application "
        f"for vacancy {vacancy_id}"
    )
    db: Session = SessionLocal()
    try:
        resume = db.query(Resume).filter(Resume.id == resume_id).first()
        if resume:
            logger.info(f"[resume-service] Resume '{resume.title}' submitted to vacancy {vacancy_id}.")
    finally:
        db.close()


async def handle_application_status_changed(payload: dict):
    """
    Fired when employer changes application status (invited/rejected/viewed).
    resume-service logs the outcome for the applicant's history.
    """
    resume_id = payload.get("resume_id")
    new_status = payload.get("status")
    vacancy_id = payload.get("vacancy_id")
    logger.info(
        f"[resume-service] Application status changed: "
        f"resume_id={resume_id}, vacancy_id={vacancy_id}, new_status={new_status}"
    )


async def start_consumer():
    """Start consuming events from RabbitMQ. Called on app startup."""
    try:
        connection = await aio_pika.connect_robust(RABBITMQ_URL)
        channel = await connection.channel()
        await channel.set_qos(prefetch_count=10)

        exchange = await channel.declare_exchange(
            EXCHANGE_NAME, aio_pika.ExchangeType.TOPIC, durable=True
        )

        queue = await channel.declare_queue(
            "resume_service_queue", durable=True
        )
        await queue.bind(exchange, routing_key="application.created")
        await queue.bind(exchange, routing_key="application.status_changed")

        logger.info("[resume-service] RabbitMQ consumer started, waiting for events...")

        async def on_message(message: aio_pika.abc.AbstractIncomingMessage):
            async with message.process():
                try:
                    payload = json.loads(message.body.decode())
                    routing_key = message.routing_key

                    if routing_key == "application.created":
                        await handle_application_created(payload)
                    elif routing_key == "application.status_changed":
                        await handle_application_status_changed(payload)
                    else:
                        logger.warning(f"[resume-service] Unknown routing key: {routing_key}")
                except Exception as e:
                    logger.error(f"[resume-service] Error processing message: {e}")

        await queue.consume(on_message)
    except Exception as e:
        logger.error(f"[resume-service] Failed to start consumer: {e}")
